tools/migration.py

```diff
@@ -0,0 +1,248 @@
+        The path parameter are additions to the path, e.g /service/id.
+        """
+        self.new.put(self.new.get_project_service_url(service) + path_param,
+                     parameters=parameters,
+                     json=data)
+
+    def get_matching_finding_id(self, finding_id):
+        """ Tries to find a matching finding in the new instance for
+        the given findings id of the old instance.
+        If no match could be found `None` is returned.
+        """
+        finding = self.get_from_old("findings-by-id", path_param=finding_id)
+        new_findings = self.get_from_new("findings",
+                                         path_param=finding["location"]["uniformPath"])
+        for new_finding in new_findings:
+            if self.match_finding(new_finding, finding):
+                return new_finding["id"]
+        return None
+
+    @staticmethod
+    def match_finding(finding1, finding2):
+        """ Checks if the given two findings are the same. """
+        if finding1["message"] != finding2["message"]:
+            return False
+
+        # some findings don't have a start line
+        has_line1 = "rawStartLine" in finding1["location"]
+        has_line2 = "rawStartLine" in finding2["location"]
+        if has_line1 != has_line2:
+            return False
+        if not has_line1:
+            return True
+        return finding1["location"]["rawStartLine"] == finding2["location"]["rawStartLine"]
+
+    @abstractmethod
+    def migrate(self):
+        pass
+
+
+class BlacklistMigrator(MigratorBase):
+    """ Class for migrating blacklists between two instances.
+    If some blacklisted finding cannot be found in the new instance, they will not
+    be migrated.
+    """
+
+    def migrate(self):
+        """ Migrates the blacklists. """
+        migrate_blacklist = self.get_blacklist_infos()
+        for blacklisted in migrate_blacklist:
+            new_id = self.get_matching_finding_id(blacklisted["findingId"])
+            if not new_id:
+                print("could not find finding %s in new Teamscale" % blacklisted["findingId"])
+            else:
+                print("mapped old finding %s to new finding %s" % (blacklisted["findingId"], new_id))
+                self.blacklist(blacklisted, new_id)
+
+        if len(migrate_blacklist) == 0:
+            print("No new blacklisted findings to migrate.")
+        else:
+            print("migrated %d/%d blacklisted findings" % (self.migrated, len(migrate_blacklist)))
+
+    def get_blacklist_infos(self):
+        """ Returns all blacklist info objects from the old instance. """
+        # Remove findings which have already been migrated and have the same id
+        blacklisted_ids = set(self.get_from_old("finding-blacklist")) - set(self.get_from_new("finding-blacklist"))
+
+        infos = []
+        for finding_id in blacklisted_ids:
+            info = self.get_from_old("finding-blacklist", path_param=finding_id)
+            if not info:
```

Comment:
are you sure this actually works? I think the server will send a 404 response (which is not valid json) which you don't handle specifically above in #get() so I think this might just throw some exception if the blacklist entry can't be found

Reply:
Well, there no reason why this shouldn't find anything tbh.
The ids I am getting the info for where retrieved with the same service call.
Calling "finding-blacklist" without arguments gives you all ids. And then I am calling the same service with the ids.
If this fails, there is something really wrong and it actually should fail